torchdata/nodes/base_node.py

```diff
@@ -5,155 +5,88 @@
 
 logger = logging.getLogger(__name__)
 
 
 T = TypeVar("T", covariant=True)
 
 
-class BaseNodeIterator(Iterator[T]):
-    def state_dict(self) -> Dict[str, Any]:
-        raise NotImplementedError()
-
-    def started(self) -> bool:
-        raise NotImplementedError()
-
-    def finished(self) -> bool:
-        raise NotImplementedError()
+class BaseNode(Iterator[T]):
+    """BaseNodes are iterators. They have the following **public** interface:
+
+    * reset(initial_state: Optional[dict] = None) - resets iterator to either initial_state or beginning if None is passed
+    * state_dict() -> Dict[str, Any] - returns a state_dict that may be passed to reset() at some point in the future
+    * __next__() -> T - users should call next(my_instance) on the iterator in order to iterate through it.
+
+    Base nodes also work in for loops as usual, if they are wrapped with an iter.
+    They can also be used directly, eg when composing BaseNodes, with a slight modification eg:
+    ```python
+    node = MyBaseNodeImpl()
+    loader = Loader(node)
+    # loader also supports state_dict() and load_state_dict()
+    for epoch in range(5):
+        for idx, batch in enumerate(loader):
+            ...
+
+    # or if using node directly:
+    node = MyBaseNodeImpl()
+    for epoch in range(5):
+        node.reset()
+        for idx, batch in enumerate(loader):
+            ...
+    ```
+
+    Subclasses of base node must implement the following methods:
+
+    * __init__() - must call super().__init__()
+    * reset(initial_state: Optional[dict]=None) - As above. Reset is a good place to put expensive
+        initialization, as it will be lazily called when next() or state_dict() is called.
+        Must call super().reset(initial_state)
+    * next() -> T - logic for returning the next value in the sequence, or throw StopIteration
+    * get_state(self) -> dict: returns a dictionary representing state that may be passed to reset()
 
+    """
 
-class BaseNode(Iterable[T]):
-    __it: Optional[BaseNodeIterator[T]] = None  # holds pointer to last iter() requested
-    __initial_state: Optional[Dict[str, Any]] = None
-    __restart_on_stop_iteration: bool = False
+    def __init__(self, *args, **kwargs):
+        self.__initialized = False
 
-    def iterator(self, initial_state: Optional[dict]) -> Iterator[T]:
-        """Override this method to implement the iterator.
-        Iterators are expected to raise StopIteration to signal
-        end of iteration, so they can be used in for loops.
-        Generators just need to return, as usual.
+    def __iter__(self):
+        return self
 
-        initial_state will be passed if `load_state_dict(initial_state)` was called
-        on this node before the __iter__ is requested, otherwise None will be passed
-        """
-        raise NotImplementedError()
+    def reset(self, initial_state: Optional[dict] = None):
+        self.__initialized = True
```

Comment:
We don't _force_ users to define reset?

Reply:
We do, in a bit of roundabout way: if users don't implement .reset() (and call super().reset()) we throw an error during next() and state_dict() calls